src/backend/cpu/sift.cpp

```diff
@@ -0,0 +1,1046 @@
+    const unsigned max_feat,
+    const float threshold)
+{
+    const af::dim4 idims = center.dims();
+    const T* prev_ptr    = prev.get();
+    const T* center_ptr  = center.get();
+    const T* next_ptr    = next.get();
+
+    for (int y = ImgBorder; y < idims[1]-ImgBorder; y++) {
+        for (int x = ImgBorder; x < idims[0]-ImgBorder; x++) {
+            float p = center_ptr[y*idims[0] + x];
+
+            // Find extrema
+            if (abs((float)p) > threshold &&
+                ((p > 0                                && p > center_ptr[(y-1)*idims[0] + x-1] && p > center_ptr[(y-1)*idims[0] + x]   &&
+                  p > center_ptr[(y-1)*idims[0] + x+1] && p > center_ptr[y*idims[0] + (x-1)]   && p > center_ptr[y*idims[0] + x+1]     &&
+                  p > center_ptr[(y+1)*idims[0] + x-1] && p > center_ptr[(y+1)*idims[0] + x]   && p > center_ptr[(y+1)*idims[0] + x+1] &&
+                  p > prev_ptr[(y-1)*idims[0] + x-1]   && p > prev_ptr[(y-1)*idims[0] + x]     && p > prev_ptr[(y-1)*idims[0] + x+1]   &&
+                  p > prev_ptr[y*idims[0] + x-1]       && p > prev_ptr[y*idims[0] + x]         && p > prev_ptr[y*idims[0] + x+1]       &&
+                  p > prev_ptr[(y+1)*idims[0] + x-1]   && p > prev_ptr[(y+1)*idims[0] + x]     && p > prev_ptr[(y+1)*idims[0] + x+1]   &&
+                  p > next_ptr[(y-1)*idims[0] + x-1]   && p > next_ptr[(y-1)*idims[0] + x]     && p > next_ptr[(y-1)*idims[0] + x+1]   &&
+                  p > next_ptr[y*idims[0] + x-1]       && p > next_ptr[y*idims[0] + x]         && p > next_ptr[y*idims[0] + x+1]       &&
+                  p > next_ptr[(y+1)*idims[0] + x-1]   && p > next_ptr[(y+1)*idims[0] + x]     && p > next_ptr[(y+1)*idims[0] + x+1])  ||
+                 (p < 0                                && p < center_ptr[(y-1)*idims[0] + x-1] && p < center_ptr[(y-1)*idims[0] + x]   &&
+                  p < center_ptr[(y-1)*idims[0] + x+1] && p < center_ptr[y*idims[0] + (x-1)]   && p < center_ptr[y*idims[0] + x+1]     &&
+                  p < center_ptr[(y+1)*idims[0] + x-1] && p < center_ptr[(y+1)*idims[0] + x]   && p < center_ptr[(y+1)*idims[0] + x+1] &&
+                  p < prev_ptr[(y-1)*idims[0] + x-1]   && p < prev_ptr[(y-1)*idims[0] + x]     && p < prev_ptr[(y-1)*idims[0] + x+1]   &&
+                  p < prev_ptr[y*idims[0] + x-1]       && p < prev_ptr[y*idims[0] + x]         && p < prev_ptr[y*idims[0] + x+1]       &&
+                  p < prev_ptr[(y+1)*idims[0] + x-1]   && p < prev_ptr[(y+1)*idims[0] + x]     && p < prev_ptr[(y+1)*idims[0] + x+1]   &&
+                  p < next_ptr[(y-1)*idims[0] + x-1]   && p < next_ptr[(y-1)*idims[0] + x]     && p < next_ptr[(y-1)*idims[0] + x+1]   &&
+                  p < next_ptr[y*idims[0] + x-1]       && p < next_ptr[y*idims[0] + x]         && p < next_ptr[y*idims[0] + x+1]       &&
+                  p < next_ptr[(y+1)*idims[0] + x-1]   && p < next_ptr[(y+1)*idims[0] + x]     && p < next_ptr[(y+1)*idims[0] + x+1]))) {
+
```

Comment:
Infact it would be better to add a macro like this:

```
#define CPTR(xx, yy) ((center_ptr[idims[0] * (yy) + (xx)))
...
...
..
p > CPTR(x+y, y-1) ...
```